#Day3 Exercise.

sanjaypradeep opened the pull request:
@pylenin please review.

Commit:
Day 2, Q1 Solution.

Forked form PyLenin - Q1 solved.

Commit:
Day 2, Q2 solutions.

Day 2, Q2 solutions.. good understanding about date difference via timedelta type.

Commit:
Addressing comments.

@pylenin, addressed the comments, expected output is obtained. seconds component is not included as of now, let me know if needed will update that as well.

Commit:
#Day3 Exercise PR

PR for two problems. @pylenin please review.

sanjaypradeep committed:
Merge branch 'master' into master

pylenin commented on .gitignore:
```diff
@@ -1,3 +1,6 @@
-
-.idea/
-.DS_Store/
+.idea/encodings.xml
```
You don't need to update .gitignore. `.idea/` will cover all the unnecessary files that are there within this folder. Please remove it.

sanjaypradeep replied:
Sure. Changes are reverted back, so resolving this conversation.

pylenin commented on Day 2 - Python Datetime objects and Timedelta/Q1_BeforeChristmas.py:
```diff
@@ -1,10 +1,10 @@
 1. What date is 3 weeks and 4 days before Christmas?
 Solution provided by Sanjay Siddha
 """
 
```
Don't add any changes here. revert these files to the way they were. Use `git pull --rebase origin master`.

pylenin commented on Day 2 - Python Datetime objects and Timedelta/Q2_TimeDifference.py:
```diff
@@ -5,7 +5,6 @@
 Solution provided by Sanjay Siddha
 """
 
-
```
same here. This is not necessary.

pylenin commented on Day 3 - Datetime strftime and strptime/Q2_Solution.py:
```diff
@@ -0,0 +1,12 @@
+
+
+inputDate = datetime(year=2019, day=20, month=1, hour=9, minute=30)
+print(datetime.strftime(inputDate, "%b %d, %Y %-H:%M %p"))
```
There is an error here. `%-H`.

sanjaypradeep replied:
Nope, it'll not throw any error, because by default `%H` will give output as `09`, the expected time output  has `9:30`, therefore to neglect `0`, so used `%-H`.

Hope this is clear.

pylenin replied:
Okay. I should have explained better. My issue is not with `-`. my issue is with the use of `%H` part.
Why?
`%H` is a 24-hour clock as a zero-padded number. When we are using AM/PM in our date time strings, we expect our clock to be a 12 hour clock. For example, `14:30 PM` doesn't really make sense. To use a 12 hour clock, you should use `%I`.
Look at the below examples.
```
# Using %H

from datetime import datetime
inputDate = datetime(year=2019, day=20, month=1, hour=14, minute=30)
print(datetime.strftime(inputDate, "%b %d, %Y %-H:%M %p"))
```
This will return 
```
Jan 20, 2019 14:30 PM
```
To me, having a 24 hour clock with AM/PM details doesn't make sense. You got lucky that the hour was 9. Hence your solution looks okay. But its not going to be the case all the time.

```
# Using %I
from datetime import datetime
inputDate = datetime(year=2019, day=20, month=1, hour=14, minute=30)
print(datetime.strftime(inputDate, "%b %d, %Y %-I:%M %p"))
```
This snippet will return you,
```
Jan 20, 2019 2:30 PM
```
This makes sense to me. 

You might think this small change that I am asking you to consider is trivial. But when it comes to working with unlabelled datasets, using the right formatting techniques are highly essential. Also anyone else who reads it, might get confused when they see `%H` being used for 12 hour clock. 

Hope this explanation is clear.

pylenin replied:
Moreover, you didn't have to do the first part. You are first creating a datetime object and converting it to a string. Then you don't use it in code. You create your own `InputString` variable. and then convert it to datetime object.

sanjaypradeep replied:
Yeah, thanks for the good explanation. Will be removing the first part and also, 12 hour clock will be included.

pylenin replied:
Actually, Mostapha has already provided the solution. Sorry about this time. but I will accept his solution for Day 3. Don't get discouraged and keep participating. You can close this PR for now. And please run a `rebase` on your repository. So you get all the changes, once I have updated.

sanjaypradeep replied:
Sure, thanks for the motivation.

Commit:
Addressing comments

Commit:
Addressing comments

Commit:
Reverting back conflicted commits

Commit:
Resolving conflicts and addressing comments

Commit:
Q2 - resolving conflicts and addressing comments

sanjaypradeep committed:
Changing same as original.

sanjaypradeep committed:
Same as original

sanjaypradeep committed:
Removing unwanted strings.

Commit:
Adding docs for the solutions.

Adding doscs/explanation about the solution. Would be helpful for someone who reads the code.

Commit:
Addressing review comments.

As discussed in the review, addressing those explained information. Thanks to @pylenin

Commit:
Merge branch 'master' of https://github.com/sanjaypradeep/Python30
